plugins/image-prioritizer/helper.php

```diff
@@ -137,14 +142,173 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+			)
+		);
+	}
+
+	/*
+	 * Validate that the Content-Length is not too massive, as it would be better to err on the side of
+	 * not preloading something so weighty in case the image won't actually end up as LCP.
+	 * The value of 2MB is chosen because according to Web Almanac 2022, the largest image by byte size
+	 * on a page is 1MB at the 90th percentile: <https://almanac.httparchive.org/en/2022/media#fig-12>.
+	 * The 2MB value is double this 1MB size.
+	 */
+	$content_length = (array) wp_remote_retrieve_header( $r, 'content-length' );
+	if ( ! is_numeric( $content_length[0] ) ) {
+		return new WP_Error(
+			'background_image_content_length_unknown',
+			__( 'HEAD request for background image URL did not include a Content-Length response header.', 'image-prioritizer' )
+		);
+	} elseif ( (int) $content_length[0] > 2 * MB_IN_BYTES ) {
+		return new WP_Error(
+			'background_image_content_length_too_large',
+			sprintf(
+				/* translators: %s is the content length of the response  */
+				__( 'HEAD request for background image URL returned Content-Length greater than 2MB: %s.', 'image-prioritizer' ),
+				$content_length[0]
+			)
+		);
+	}
+
+	return true;
+}
+
+/**
+ * Filters the validity of a URL Metric with an LCP element background image.
+ *
+ * This removes the lcpElementExternalBackgroundImage from the URL Metric prior to it being stored if the background
+ * image URL is not valid. Removal of the property is preferable to
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param bool|WP_Error|mixed  $validity   Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ * @param OD_Strict_URL_Metric $url_metric URL Metric, already validated against the JSON Schema.
+ * @return bool|WP_Error Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ *
+ * @noinspection PhpDocMissingThrowsInspection
+ * @throws OD_Data_Validation_Exception Except it won't because lcpElementExternalBackgroundImage is not a required property.
+ */
+function image_prioritizer_filter_store_url_metric_validity( $validity, OD_Strict_URL_Metric $url_metric ) {
+	if ( ! is_bool( $validity ) && ! ( $validity instanceof WP_Error ) ) {
+		$validity = (bool) $validity;
+	}
+
+	$data = $url_metric->get( 'lcpElementExternalBackgroundImage' );
+	if ( is_array( $data ) && isset( $data['url'] ) && is_string( $data['url'] ) ) { // Note: The isset() and is_string() checks aren't necessary since the JSON Schema enforces them to be set.
+		$image_validity = image_prioritizer_validate_background_image_url( $data['url'] );
+		if ( is_wp_error( $image_validity ) ) {
+			/**
+			 * No WP_Exception is thrown by wp_trigger_error() since E_USER_ERROR is not passed as the error level.
+			 *
+			 * @noinspection PhpUnhandledExceptionInspection
+			 */
+			wp_trigger_error( __FUNCTION__, $image_validity->get_error_message() . ' Background image URL: ' . $data['url'] );
+			$url_metric->unset( 'lcpElementExternalBackgroundImage' );
```

Comment:
My thinking behind unset here is that what if someone using an external background image from another domain not in the allowed list of origins. If any exception it thrown then every single URL Metric will be rejected. This is not good because then none of the other data in the URL Metric would ever be stored. What we want in this case is to just exclude that non-required property from the URL Metric as if it wasn't set in the first place. The use of `wp_trigger_error()` is to provide a mechanism for a site owner to find out when `WP_DEBUG` is enabled.

Reply:
That makes sense, but then shouldn't we have a filter for sanitization instead for this purpose, rather than validation?

Given this is for now the only usage of the validation filter, but it's not throwing an exception, I think it would make sense to change this to a sanitization filter that acts on the `URL_Metric` instance. This would be similar to e.g. how Core has a sanitization filter for options, but not validation. Obviously we could also add a validation extension point, but that could then happen later and doesn't seem as relevant to this as a sanitization filter.

Reply:
Here's an alternative to the current validity filter in this PR. We could have a filter that runs on the `$data` before it is passed into the `OD_Strict_URL_Metric` class for instantiation. So instead of:

https://github.com/WordPress/performance/blob/6005f4a9f084825edc4f2eeeb4152b41b29ef0ce/plugins/optimization-detective/storage/rest-api.php#L188-L200

We could have:

```php
try {
	$data = array_merge(
		$data,
		array(
			// Now supply the readonly args which were omitted from the REST API params due to being `readonly`.
			'timestamp' => microtime( true ),
			'uuid'      => wp_generate_uuid4(),
			'etag'      => $request->get_param( 'current_etag' ),
		)
	);

	/**
	 * Filters the URL Metric data prior to validation for storage.
	 *
	 * This can be used to apply additional sanitization to the URL Metric data which
	 * cannot be done with JSON Schema alone. Additionally, this filter can be used to
	 * apply validation constraints by throwing an OD_Data_Validation_Exception which
	 * will block the URL Metric from being stored.
	 *
	 * @since n.e.x.t
	 * @param array<string, mixed> $data URL Metric data. This is the Data type from OD_URL_Metric.
	 */
	$data = apply_filters( 'od_url_metric_data_pre_storage', $data );

	// The "strict" URL Metric class is being used here to ensure additionalProperties of all objects are disallowed.
	$url_metric = new OD_Strict_URL_Metric( $data );
```

In this way, the same filter can be used for both sanitization and validation. Even without explicitly throwing an exception, an exception would end up getting thrown anyway if a filter did something like `unset( $data['viewport] )` since that is a required property.

Reply:
See 94c5e6b

Reply:
@westonruter That looks good to me. Though I think that filter name would be better as `od_sanitize_url_metric_data`, as it's a more intuitive name in terms of naming structure, and its primary purpose would be to sanitize data.

Makes sense that this filter could also be used for validation, so the `try catch` wrap around it makes sense.

Reply:
I don't think using "sanitize" in the filter name makes sense since other core filters aren't named in that way. It's up to the filter callback to decide whether the purpose is to sanitize. Going the Options API as a model, there is the `pre_update_option` filter which is used to (maybe) serialize, how about `od_pre_store_url_metric`? Or to follow `wp_insert_post()` which has the `wp_insert_post_data` filter, what about `od_store_url_metric_data`? I like that more.

Reply:
> I don't think using "sanitize" in the filter name makes sense since other core filters aren't named in that way.

I don't think that's true, see e.g. `sanitize_option`.

> It's up to the filter callback to decide whether the purpose is to sanitize. Going the Options API as a model, there is the `pre_update_option` filter which is used to (maybe) serialize, how about `od_pre_store_url_metric`?

What other use-cases would such a filter have than sanitize? You can throw an exception (effectively validate), but technically you can throw an exception from anywhere, so that to me is not necessarily a reason to change the name. The example of serialization wouldn't be relevant here.

> Or to follow `wp_insert_post()` which has the `wp_insert_post_data` filter, what about `od_store_url_metric_data`? I like that more.

This would be an okay name to me if we go with filtering the raw data. But that's not as clean of an API surface as filtering the metric instance (per my other comment), so I would prefer not to do that.

Reply:
> What other use-cases would such a filter have than sanitize? You can throw an exception (effectively validate), but technically you can throw an exception from anywhere, so that to me is not necessarily a reason to change the name. The example of serialization wouldn't be relevant here.

A use case could be an extension which amends a stored URL Metric with additional data, like the IP address of the client which submitted the URL Metric.

> This would be an okay name to me if we go with filtering the raw data. But that's not as clean of an API surface as filtering the metric instance (per my other comment), so I would prefer not to do that.

I think the API surface is actually cleaner by passing around an array rather then the `OD_URL_Metric` instance. See https://github.com/WordPress/performance/pull/1713#discussion_r1887354549.